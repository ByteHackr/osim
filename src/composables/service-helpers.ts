import { getDisplayedOsidbError } from '@/services/OsidbAuthService';
import { useToastStore } from '@/stores/ToastStore';
import { AxiosError, type AxiosResponse } from 'axios';

export function createCatchHandler(title: string = 'Error', shouldThrow: boolean = true) {
  return (error: AxiosError) => {
    const displayedError = getDisplayedOsidbError(error);
    const { addToast } = useToastStore();
    addToast({
      title,
      body: `${displayedError}: ${error.response?.data}`,
      css: 'warning',
    });
    console.error('‚ùå ', error);
    if (shouldThrow) throw error;
    return error;
  };
}

export function createSuccessHandler({ title, body }: Record<string, string>, onSucceed: Function = () => {}) {
  return (response: AxiosResponse) => {
    const { addToast } = useToastStore();
    onSucceed(response);
    addToast({
      title,
      body,
      css: 'success',
    });
  };
}
