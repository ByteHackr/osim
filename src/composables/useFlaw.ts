import { ref, readonly, computed } from 'vue';

import type { ZodFlawType } from '@/types/zodFlaw';
import { deepCopyFromRaw, serializeWithExclude } from '@/utils/helpers';

const flaw = ref<ZodFlawType>(blankFlaw());
const initialFlaw = ref<ZodFlawType>(blankFlaw());
const relatedFlaws = ref<ZodFlawType[]>([]);

function resetFlaw() {
  flaw.value = blankFlaw();
  initialFlaw.value = blankFlaw();
  relatedFlaws.value = [];
}

function setFlaw(fetchedFlaw: ZodFlawType) {
  initialFlaw.value = deepCopyFromRaw(fetchedFlaw);
  flaw.value = fetchedFlaw;
}

const isFlawUpdated = computed(
  () => {
    const keysToExclude: Array<keyof ZodFlawType> = ['affects', 'trackers', 'cvss_scores'];
    return serializeWithExclude(flaw.value, keysToExclude)
      !== serializeWithExclude(initialFlaw.value, keysToExclude);
  },
);

export function useFlaw() {
  return { flaw, initialFlaw: readonly(initialFlaw), isFlawUpdated, relatedFlaws, resetFlaw, setFlaw };
}

export function blankFlaw(): ZodFlawType {
  return {
    affects: [],
    classification: {
      state: 'NEW',
      workflow: '',
    },
    components: [],
    unembargo_dt: null,
    reported_dt: new Date().toISOString(),
    uuid: '',
    cve_id: '',
    cvss_scores: [],
    cwe_id: '',
    comment_zero: '',
    embargoed: false,
    impact: null,
    major_incident_state: '',
    source: '',
    title: '',
    owner: '',
    team_id: '',
    cve_description: '',
    statement: '',
    mitigation: '',
    task_key: '',
    comments: [],
    labels: [],
    references: [],
    acknowledgments: [],
    alerts: [],
  };
}
