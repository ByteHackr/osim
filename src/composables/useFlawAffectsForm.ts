import { type Ref, ref, toRef, computed } from 'vue';

import { type Flaw, type Affect } from '@/generated-client';
import { postAffect, putAffect } from '@/services/AffectService';
import { getDisplayedOsidbError } from '@/services/OsidbAuthService';
import { useToastStore } from '@/stores/ToastStore';

export function useFlawAffectsForm(flaw: Ref<Flaw>) {
  const wereAffectsModified = ref(false);
  const modifiedAffectIds = ref<string[]>([]);
  const affectIdsToDelete = ref<string[]>([]);
  const theAffects = toRef(flaw.value, 'affects') as Ref<Affect[]>;

  const affectsToSave = computed(() => [
    ...theAffects.value.filter((affect) => modifiedAffectIds.value.includes(affect.uuid)),
    ...theAffects.value.filter((affect) => !affect.uuid),
  ]);

  const { addToast } = useToastStore();

  function addBlankAffect() {
    theAffects.value.push({} as Affect);
  }

  function removeAffect(affectIdx: number) {
    // TODO: Send delete request to OSIDB
    affectIdsToDelete.value.push(theAffects.value[affectIdx].uuid);
    theAffects.value.splice(affectIdx, 1);
  }

  function reportAffectAsModified(affectId: string) {
    wereAffectsModified.value = true;
    modifiedAffectIds.value.push(affectId);
    console.log(modifiedAffectIds.value);
  }

  const saveAffects = async () => {
    const affectsToSaveQuantity = affectsToSave.value.length;
    for (let [index, affect] of Object.entries(affectsToSave.value)) {
      console.log(`Saving affect ${+index + 1} of ${affectsToSaveQuantity}`, affect);
      console.log(affect.uuid);
      const requestBody = {
        flaw: flaw.value?.uuid,
        type: affect.type,
        affectedness: affect.affectedness,
        resolution: affect.resolution,
        ps_module: affect.ps_module,
        ps_component: affect.ps_component,
        impact: affect.impact,
        embargoed: affect.embargoed || false,
        updated_dt: affect.updated_dt,
      };
      if (affect.uuid != null) {
        await putAffect(affect.uuid, requestBody)
          .then(() => {
            console.log('saved newAffect', requestBody);
            addToast({
              title: 'Info',
              body: `Affect ${+index + 1} of ${affectsToSaveQuantity} Saved: ${
                requestBody.ps_component
              }`,
            });
          })
          .catch((error) => {
            const displayedError = getDisplayedOsidbError(error);
            addToast({
              title: 'Error saving Affect',
              body: displayedError,
            });
            console.log(error);
          });
      } else {
        await postAffect(requestBody)
          .then(() => {
            console.log('saved newAffect', requestBody);
            addToast({
              title: 'Info',
              body: `Affect Saved: ${requestBody.ps_component}`,
            });
          })
          .catch((error) => {
            const displayedError = getDisplayedOsidbError(error);
            addToast({
              title: 'Error saving Affect',
              body: displayedError,
            });
            console.log(error);
          });
      }
    }
  };

  return {
    wereAffectsModified,
    addBlankAffect,
    removeAffect,
    saveAffects,
    theAffects,
    reportAffectAsModified,
  };
}

