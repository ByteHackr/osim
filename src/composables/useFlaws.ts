import { ref } from 'vue';
import { getFlaws } from '@/services/FlawService';

export function useFlaws() {
  const isFinalPageFetched = ref(false);
  const isLoading = ref(false);
  const issues = ref<any[]>([]);
  const offset = ref(0);
  const pagesize = 20;
  const total = ref(0);

  function loadFlaws(params: any = {}) {
    offset.value = 0;
    isFinalPageFetched.value = false;  
    isLoading.value = true;
    issues.value = [];
    total.value = 0;

    getFlaws(offset.value, 100, params.value)
      .then((response) => {
        issues.value = response.data.results;
        // response will have next property for next api call
        if (response.data.next) {
          const url = new URL(response.data.next);
          const params = new URLSearchParams(url.search);
          offset.value = Number(params.get('offset') || 0);
        } else {
          isFinalPageFetched.value = true;
          offset.value += response.data.results.length; 
        }

        total.value = response.data.count;
      })
      .catch((err) => {
        console.error('IssueQueue: getFlaws error: ', err);
      })
      .finally(() => {
        isLoading.value = false;
      });
  }

  function loadMoreFlaws(params: any = {}) {
    if (isLoading.value || isFinalPageFetched.value) {
      return; // Early exit if already loading
    }
    isLoading.value = true;

    getFlaws(offset.value, pagesize, params.value)
      .then((response) => {
        issues.value = [...issues.value, ...response.data.results];
        // response will have next property for next api call
        if (response.data.next) {
          const url = new URL(response.data.next);
          const params = new URLSearchParams(url.search);
          offset.value = Number(params.get('offset') || 0);
        } else {
          isFinalPageFetched.value = true;
          offset.value += response.data.results.length; 
        }
      })
      .catch((err) => {
        console.error('Error fetching more flaws: ', err);
      })
      .finally(() => {
        isLoading.value = false;
      });
  }

  return {
    issues,
    isFinalPageFetched,
    isLoading,
    offset,
    pagesize,
    total,
    loadFlaws,
    loadMoreFlaws
  };
}
