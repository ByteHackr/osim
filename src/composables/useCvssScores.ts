import { putFlawCvssScores, postFlawCvssScores } from '@/services/FlawService';
import { type Flaw } from '@/generated-client';
import { createSuccessHandler, createCatchHandler } from './service-helpers';
import { computed, ref, watch, type Ref } from 'vue';

const onSucceed = createSuccessHandler({ title: 'Success!', body: 'Saved CVSS Scores' });
const onError = createCatchHandler('Error updating Flaw CVSS data');

export function useCvssScores(flaw: Ref<Flaw>) {
  const wasCvssModified = ref(false);
  const redHatCvssData = flaw.value.cvss_scores.find((assessment) => assessment.issuer === 'RH');
  const flawRhCvss = ref(redHatCvssData || flaw.value.cvss_scores[0]);
  
  watch(flawRhCvss, () => wasCvssModified.value = true, {deep: true});

  const nistCvssScore = computed(() => flaw.value.cvss_scores.find((score) => score.issuer === 'NIST'));
  const flawNvdCvssScore = computed(() => nistCvssScore.value?.score || flaw.value.nvd_cvss3);

  async function saveCvssScores() {
    if (flawRhCvss.value.created_dt) {
      return putFlawCvssScores(
        flaw.value.uuid,
        flawRhCvss.value.uuid || '',
        flawRhCvss.value as unknown
      )
        .then(onSucceed)
        .catch(onError);
    } else {
      const requestBody = {
        // "score":  is recalculated based on the vector by OSIDB and does not need to be included
        comment: flawRhCvss.value.comment,
        cvss_version: 'V3',
        issuer: 'RH',
        vector: flawRhCvss.value.vector,
        embargoed: flaw.value.embargoed,
      };
      return postFlawCvssScores(flaw.value.uuid, requestBody as unknown)
        .then(onSucceed)
        .catch(onError);
    }
  }

  return {
    wasCvssModified,
    flawNvdCvssScore,
    flawRhCvss,
    saveCvssScores,
  };
}

