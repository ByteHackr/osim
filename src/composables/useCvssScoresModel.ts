import { putFlawCvssScores, postFlawCvssScores } from '@/services/FlawService';
import { createSuccessHandler, createCatchHandler } from './service-helpers';
import { computed, ref, watch, type Ref } from 'vue';
import type { ZodFlawType } from '@/types/zodFlaw';

const onSucceed = createSuccessHandler({ title: 'Success!', body: 'Saved CVSS Scores' });
const onError = createCatchHandler('Error updating Flaw CVSS data');

export function useCvssScoresModel(flaw: Ref<ZodFlawType>) {
  const wasCvssModified = ref(false);
  const redHatCvssData = flaw.value.cvss_scores.find((assessment) => assessment.issuer === 'RH');
  const flawRhCvss = ref(
    redHatCvssData || {
      vector: null,
      uuid: null,
      issuer: null,
      comment: null,
      created_dt: null,
      score: null,
    }, // empty for creating a flaw, or if no CVSS data exists
  );

  watch(flawRhCvss, () => (wasCvssModified.value = true), { deep: true });

  const nistCvssScore = computed(() =>
    flaw.value.cvss_scores.find((score) => score.issuer === 'NIST'),
  );

  const nvdCvssString = computed(() => {
    const values = [nistCvssScore.value?.score, nistCvssScore.value?.vector].filter(Boolean);
    return values.join('/') || flaw.value.nvd_cvss3 || '-';
  });

  const cvssString = computed(() => {
    const values = [flawRhCvss.value?.score, flawRhCvss.value?.vector].filter(Boolean);
    return values.join('/');
  });

  const shouldDisplayEmailNistForm = computed(() => {
    if(cvssString.value === '' || nvdCvssString.value === '-') {
      return false;
    }
    return `${cvssString.value}` !== `${nvdCvssString.value}`;
  });

  const highlightedNvdCvssScore = computed(() => {
    const result = [];
    const nvdCvssValue = nvdCvssString.value;
    const cvssValue = cvssString.value;
    const maxLength = Math.max(nvdCvssValue.length, cvssValue.length);

    for (let i = 0; i < maxLength; i++) {
      const charFromFlaw = i < nvdCvssValue.length ? nvdCvssValue[i] : '';
      const charFromCvss = i < cvssValue.length ? cvssValue[i] : '';
      result.push({ 
        char: charFromFlaw, 
        isHighlighted: shouldDisplayEmailNistForm.value && charFromFlaw !== charFromCvss,
      });
    }

    return result;
  });

  async function saveCvssScores() {
    if (flawRhCvss.value.created_dt) {
      return putFlawCvssScores(
        flaw.value.uuid,
        flawRhCvss.value.uuid || '',
        flawRhCvss.value as unknown,
      )
        .then(onSucceed)
        .catch(onError);
    }

    const requestBody = {
      // "score":  is recalculated based on the vector by OSIDB and does not need to be included
      comment: flawRhCvss.value.comment,
      cvss_version: 'V3',
      issuer: 'RH',
      vector: flawRhCvss.value.vector,
      embargoed: flaw.value.embargoed,
    };
    return postFlawCvssScores(flaw.value.uuid, requestBody as unknown)
      .then(onSucceed)
      .catch(onError);
  }

  return {
    wasCvssModified,
    flawRhCvss,
    shouldDisplayEmailNistForm,
    highlightedNvdCvssScore,
    cvssString,
    nvdCvssString,
    saveCvssScores,
  };
}
