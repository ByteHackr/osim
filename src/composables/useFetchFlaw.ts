import { ref } from 'vue';

import { resetInitialAffects } from '@/composables/useFlawAffectsModel';

import { getFlaw, getRelatedFlaws } from '@/services/FlawService';
import type { ZodAffectType, ZodFlawType } from '@/types';
import { useToastStore } from '@/stores/ToastStore';
import { getDisplayedOsidbError } from '@/services/osidb-errors-helpers';

const flaw = ref<null | ZodFlawType>(null);
const relatedFlaws = ref<ZodFlawType[]>([]);
const isFetchingRelatedFlaws = ref(false);

export function initializeFlaw() {
  flaw.value = null;
  relatedFlaws.value = [];
  resetInitialAffects();
}

export function useFetchFlaw() {
  const { addToast } = useToastStore();

  const didFetchFail = ref<boolean>(false);

  async function fetchRelatedFlaws(affects: ZodAffectType[]) {
    isFetchingRelatedFlaws.value = true;
    try {
      relatedFlaws.value = await getRelatedFlaws(affects);
    } catch (error) {
      console.error('useFetchRelatedFlaws::fetchRelatedFlaws()', error);
      didFetchFail.value = true;
      throw error;
    } finally {
      isFetchingRelatedFlaws.value = false;
    }
  }

  async function fetchFlaw(flawCveOrId: string) {
    try {
      const fetchedFlaw = await getFlaw(flawCveOrId);
      resetInitialAffects();
      flaw.value = Object.assign({}, fetchedFlaw);
      await fetchRelatedFlaws(fetchedFlaw.affects);
      history.replaceState(null, '', `/flaws/${(fetchedFlaw.cve_id || fetchedFlaw.uuid)}`);
    } catch (error) {
      console.error('useFetchFlaw::fetchFlaw()', error);
      didFetchFail.value = true;
      flaw.value = null;
      resetInitialAffects();
      addToast({
        title: 'Error loading Flaw',
        body: getDisplayedOsidbError(error),
      });
      console.error('FlawEditView::refreshFlaw() Error loading flaw', error);
    }
  }

  return {
    relatedFlaws,
    isFetchingRelatedFlaws,
    fetchRelatedFlaws,
    flaw,
    fetchFlaw,
    didFetchFail,
  };
}
