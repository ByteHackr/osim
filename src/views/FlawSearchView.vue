<script setup lang="ts">
import { computed, ref, toRaw, watch, type Ref } from 'vue';

import { equals } from 'ramda';

import IssueSearchAdvanced from '@/components/IssueSearchAdvanced.vue';
import IssueQueue from '@/components/IssueQueue.vue';

import { useFlawsFetching } from '@/composables/useFlawsFetching';
import { useSearchParams } from '@/composables/useSearchParams';

import { useSearchStore } from '@/stores/SearchStore';
import { useToastStore } from '@/stores/ToastStore';
import { allowedEmptyFieldMapping } from '@/constants/flawFields';

const { isFinalPageFetched, isLoading, issues, loadFlaws, loadMoreFlaws, total } = useFlawsFetching();
const { facets, getSearchParams, loadAdvancedSearch, query } = useSearchParams();

const searchStore = useSearchStore();
const { addToast } = useToastStore();
const tableFilters = ref<Record<string, string>>({});

const loadedSearch = ref<number>(-1);
const draftQuery = ref('');
const draftFields = ref({});

const facetsParsed = computed(() => facets.value.reduce(
  (fields, { field, value }) => {
    if (field && (value || allowedEmptyFieldMapping[field])) {
      fields[field] = value;
    }
    return fields;
  },
  {} as Record<string, string>,
),
);

const changedSlot = computed(() =>
  loadedSearch.value >= 0
  && (!equals(toRaw(searchStore.savedSearches[loadedSearch.value].searchFilters), facetsParsed.value)
  || !equals(searchStore.savedSearches[loadedSearch.value].queryFilter, query.value)
  ),
);

const params = computed(() => {
  const searchParams = getSearchParams();
  const order = [searchParams.order, tableFilters.value.order]
    .filter(Boolean).join(',');

  if (searchParams.order && tableFilters.value.order) {
    searchParams.order += ',' + tableFilters.value.order;
  }
  const paramsObj = {
    ...tableFilters.value,
    ...searchParams,
    order,
  };

  return paramsObj;
});

watch(() => params, () => {
  loadFlaws(params);
}, {
  deep: true,
});

function fetchMoreFlaws() {
  loadMoreFlaws(params);
}

function setTableFilters(newFilters: Ref<Record<string, string>>) {
  tableFilters.value = {
    ...newFilters.value,
  };
}

function selectSavedSearch(index: number) {
  if (loadedSearch.value === index) {
    deselectSavedSearch();
    return;
  }

  if (loadedSearch.value === -1) {
    draftQuery.value = query.value;
    draftFields.value = facets.value;
  }

  const savedQuery = searchStore.savedSearches.at(index)?.queryFilter || '';
  const savedFields = searchStore.savedSearches.at(index)?.searchFilters || {};
  loadAdvancedSearch(savedQuery, savedFields);
  loadedSearch.value = index;
}

function deselectSavedSearch() {
  loadAdvancedSearch(draftQuery.value, draftFields.value);
  draftQuery.value = '';
  draftFields.value = {};
  loadedSearch.value = -1;
}

function saveSearch(name: string) {
  searchStore.saveSearch(name, facetsParsed.value, query.value);
  loadedSearch.value = searchStore.savedSearches.length - 1;
  addToast({
    title: 'Search saved',
    body: 'User\'s search saved on Slot ' + (searchStore.savedSearches.length),
  });
}

function updateSavedSearch() {
  searchStore.savedSearches[loadedSearch.value].queryFilter = query.value;
  searchStore.savedSearches[loadedSearch.value].searchFilters = facetsParsed.value;
  loadAdvancedSearch(query.value, facetsParsed.value);
  addToast({
    title: 'Slot updated',
    body: 'User\'s saved Slot ' + (loadedSearch.value + 1) + ' updated',
  });
}

function deleteSavedSearch() {
  searchStore.savedSearches.splice(loadedSearch.value, 1);
  loadedSearch.value = searchStore.savedSearches.length > 0
    ? loadedSearch.value > 0
      ? loadedSearch.value - 1
      : loadedSearch.value + 1
    : -1;
  deselectSavedSearch();
}
</script>

<template>
  <main class="mt-3">
    <IssueSearchAdvanced
      :isLoading="isLoading"
      :loadedSearch
      :savedSearches="searchStore.savedSearches"
      :changedSlot
      @savedSearch:select="selectSavedSearch"
      @filter:save="saveSearch"
      @filter:update="updateSavedSearch"
      @filter:delete="deleteSavedSearch"
    />
    <IssueQueue
      :issues="issues"
      :isLoading="isLoading"
      :isFinalPageFetched="isFinalPageFetched"
      :total="total"
      @flaws:fetch="setTableFilters"
      @flaws:load-more="fetchMoreFlaws"
    />
  </main>
</template>
