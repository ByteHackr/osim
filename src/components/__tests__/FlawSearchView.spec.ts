import { mount, VueWrapper, flushPromises } from '@vue/test-utils';
import { describe, it, expect, vi } from 'vitest';
import { useRoute, useRouter } from 'vue-router';
import { createTestingPinia } from '@pinia/testing';
import FlawSearchView from '@/views/FlawSearchView.vue';
import { useFlaws }  from '../../composables/useFlaws';

vi.mock('@vueuse/core', () => ({
  useLocalStorage: vi.fn((key: string, defaults) => {
    return {
      UserStore: {
        value: defaults || {
          // Set your fake user data here
          refresh: 'mocked_refresh_token',
          env: 'mocked_env',
          whoami: {
            email: 'test@example.com',
            username: 'testuser',
          },
        },
      },
    }[key];
  }),
  useStorage: vi.fn((key: string, defaults) => {
    return {
      'OSIM::API-KEYS': {
        value: defaults || {
          bugzillaApiKey: '',
          jiraApiKey: '',
          showNotifications: false,
        },
      },
    }[key];
  }),
}));

vi.mock('jwt-decode', () => ({
  default: vi.fn(() => ({
    sub: '1234567890',
    name: 'Test User',
    exp: Math.floor(Date.now() / 1000) + 60 * 60 * 24 * 365,
  })),
}));

vi.mock('vue-router', async () => {
  const actual = await vi.importActual('vue-router');
  const replaceMock = vi.fn();
  
  return {
    ...actual,
    useRoute: vi.fn(() => ({ query: { mode: 'advanced', query: 'search' } })),
    useRouter: vi.fn(() => ({
      replace: replaceMock
    }))
  };
});

vi.mock('../../composables/useFlaws',  () => ({
  useFlaws: vi.fn(() => ({
    issues: [],
    isLoading: false,
    isFinalPageFetched: false,
    loadFlaws: vi.fn(),
    loadMoreFlaws: vi.fn(),
  })),
}));

describe('FlawSearchView', () => {
  let wrapper: VueWrapper<any>;
  const props: typeof FlawSearchView.props = {};

  beforeEach(() => {
    vi.mocked(useFlaws).mockReturnValue({
      issues: [],
      isLoading: false,
      isFinalPageFetched: false,
      loadFlaws: vi.fn(),
      loadMoreFlaws: vi.fn(),
    });
    const pinia = createTestingPinia({
      createSpy: vitest.fn,
      stubActions: false,
    });
    wrapper = mount(FlawSearchView, {
      props,
      global: {
        mocks: { useRoute },
        plugins:[pinia]
      }
    });
  });

  afterAll(() => {
    vi.clearAllMocks();
  });

  it('should render', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('should call Load on mounted', async () => {
    await flushPromises();
    expect(useFlaws().loadFlaws).toHaveBeenCalledOnce();
    expect(useFlaws().loadFlaws.mock.calls[0][0]._value).toStrictEqual({
      'order': '-created_dt',
      'search': 'search',
    });
  });

  it('should call loadFlaws on search', async () => {
    await flushPromises();
    expect(useFlaws().loadFlaws).toHaveBeenCalledOnce();
    const selectDropdown = wrapper.find('select.form-select.search-facet-field');
    await selectDropdown.setValue(selectDropdown.findAll('option')[1].element.value);
    await selectDropdown.trigger('change');
    const inputField = wrapper.find('input.form-control');
    await inputField.setValue('test'); 
    const searchButton = wrapper.find('button[type="submit"]');
    expect(searchButton.exists()).toBeTruthy();
    await searchButton.trigger('submit');
    await flushPromises();
    expect(useRouter().replace).toHaveBeenCalled();
    expect(useRouter().replace.mock.calls[0][0])
      .toStrictEqual({ query: { query: 'search', acknowledgments__name: 'test' } });
  });
});
