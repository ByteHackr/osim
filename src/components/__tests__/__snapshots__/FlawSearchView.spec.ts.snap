// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flawSearchView > should render 1`] = `
"<main class="mt-3">
  <issue-search-advanced-stub isloading="false"></issue-search-advanced-stub>
  <issue-queue-stub issues="" total="1337" isfinalpagefetched="false" isloading="false" showfilter="false" isdefaultfilterselected="true"></issue-queue-stub>
</main>"
`;
